Mathlib/Computability/NFA.lean

```diff
@@ -115,6 +123,209 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [← toDFA_correct] at hx ⊢
   exact M.toDFA.pumping_lemma hx hlen
 
+section Closure
+
+/-! ### NFA Closure Properties:
+    Properties relating operations on NFAs to their
+    languages. -/
+
+section Reversal
+
+section Auxilary
+
+/-- `M.unstep` is the reverse of `M.step`. -/
+def unstep (s : σ) (a : α) : Set σ := {s' | s ∈ M.step s' a}
+
+theorem mem_unstep {s t : σ} {a : α} : s ∈ M.unstep t a ↔ t ∈ M.step s a := Iff.rfl
+
+/-- Reversed analog of `M.stepSet S a`:
+  `M.unstepSet S a` is the union of `unstep M s a` for all `s ∈ S`.
+  It computes the set of states that have a transition in `M`
+  to a state in `S`. -/
+def unstepSet (S : Set σ) (a : α) : Set σ := ⋃ s ∈ S, M.unstep s a
+
+theorem mem_unstepSet {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, s ∈ M.unstep t a := by
+  simp [unstepSet]
+
+theorem mem_unstepSet_step {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, t ∈ M.step s a := by
+  simp [mem_unstepSet, mem_unstep]
+
+/-- Reseversed analog of `M.evalFrom S x`:
+  `M.rewindFrom S x` computes all possible reversed paths through `M` with
+  input `x` starting at an element of `S`. -/
+def rewindFrom : Set σ → List α → Set σ := List.foldl M.unstepSet
+
+/-- `M.rewind x` computes all possible paths through `M` with input `x` ending at an element of
+  `M.accept`. -/
+def rewind : List α → Set σ := M.rewindFrom M.accept
```

Comment:
This comment doesn't seem very accurate.

Reply:
Ah, sorry, I confused myself :sweat_smile:. I think made it consistent with the comment for `rewindFrom`.